mindmeld/components/_util.py

```diff
@@ -38,3 +39,207 @@ def _get_module_or_attr(module_name: str, func_name: str = None):
     if func_name not in dir(m):
         raise ImportError(f"Cannot import {func_name} from {module_name}")
     return getattr(m, func_name)
+
+
+class TreeNode:
+    def __init__(self, value: str, parent=None, children=None, allow=None):
+        self.value = value
+        self.allow = allow
+        self.parent = parent
+        self.children = children
+
+
+class TreeNlp:
+    def __init__(self, nlp, allow=None):
+        # root
+        self.root = TreeNode('root', allow=allow, children=[])
+        # construct NLP tree
+        for domain in nlp.domains:
+            domain_node = TreeNode(domain, parent=self.root, allow=allow, children=[])
+            self.root.children.append(domain_node)
+            for intent in nlp.domains[domain].intents:
+                intent_node = TreeNode(intent, parent=domain_node, allow=allow, children=[])
+                domain_node.children.append(intent_node)
+                entities = nlp.domains[domain].intents[intent].entities
+                for entity in entities:
+                    entity_node = TreeNode(entity, parent=intent_node, allow=allow, children=[])
+                    intent_node.children.append(entity_node)
+                    for role in entities[entity].role_classifier.roles:
+                        role_node = TreeNode(role, parent=intent_node, allow=allow, children=None)
+                        entity_node.children.append(role_node)
+
+    def get_domains(self):
+        return self.root.children or []
```

Comment:
To clarify, the only time the empty list would be returned would be when `self.root.children` is None, or in other words, where there aren't any domains in NLP. However, in this case should we instead raise an exception in the __init__ and prevent the creation of TreeNLP objects without any domain nodes?

Reply:
done